### ISSUE TITLE
Mem leak when queue has no consumer and tasks keep on piling up

### ISSUE THREAD

**Issue by NevermindNilas** [python/cpython#120610]
# Bug report

### Bug description:

Hi, 

In this use case I am benchmarking some decoding options.

What I Noticed that the regular `put` function of the `queue` library doesn't seem to block past the set maxsize and it instead keeps on piling up more and more tasks even if there is no consumer or the consumer can't keep up.

```python
import torch
import cv2
import decord
from tqdm import tqdm
from queue import Queue

decord.bridge.set_bridge("torch")

video = cv2.VideoCapture(r"G:\TheAnimeScripter\input\input.mp4")
totalFrameCount = video.get(cv2.CAP_PROP_FRAME_COUNT)
video.release()

queue = Queue(maxsize=100)
video = decord.VideoReader(r"G:\TheAnimeScripter\input\input.mp4", ctx=decord.cpu(0))

for i in tqdm(range(int(totalFrameCount))):
    frame = video[i].permute(2, 0, 1)
    queue.put(frame)
    #queue.get()

```

Test with `queue.get()` commented out:
![image](https://github.com/python/cpython/assets/128264457/2bc4e429-4dd1-44b9-a014-c1fea5f31f59)

Test with `queue.get()` uncommented:
![image](https://github.com/python/cpython/assets/128264457/caf33ed7-ba8c-4fbc-a618-f73603d19d0e)
( ignore the big spike, that's from the run prior to this ) 

In both cases maxsize was set to 100.
I have also tested with `block=True` but I didn't see any difference.

### CPython versions tested on:

3.11

### Operating systems tested on:

Windows

**Comment by gaogaotiantian** [comment]
In your first screenshot `tqdm` shows the frame number `70`, which is less than the maximum number set (`100`). So yes it's piling up as it does not reach the max yet. Try changing the `maxsize` to `1` and see if this still happens.

**Comment by NevermindNilas** [comment]
Memory usage with maxsize set to `1` and `no` consumer:
![image](https://github.com/python/cpython/assets/128264457/32c549d2-10cf-4c93-918f-aacd460426e2)

Memory usage with maxsize set to `1` and `with` a consumer:
![image](https://github.com/python/cpython/assets/128264457/8a90ead0-15cc-4f9d-8b30-0e221874e1e1)

Also a small sidenote that 70 was counterintuitive due to how tqdm's printing works, it's not really real time and it was lagging a bit behind. With a couple hundred it/s as seen a bit more to the right the counter should've been way beyond 70 ( at least a little over a thousand or two ) but the maxsize blocked it from growing.

For whatever it's worth, the input video I am testing with is 1920x1080p and rgb. 

Thank you.

**Comment by gaogaotiantian** [comment]
Okay I'm pretty sure this has nothing to do with CPython. I think `decord` is probably doing something in the background which eats all the memory. 

Can you try:

```python
import torch
import cv2
import decord
import time
from tqdm import tqdm

decord.bridge.set_bridge("torch")

video = cv2.VideoCapture(r"G:\TheAnimeScripter\input\input.mp4")
totalFrameCount = video.get(cv2.CAP_PROP_FRAME_COUNT)
video.release()

video = decord.VideoReader(r"G:\TheAnimeScripter\input\input.mp4", ctx=decord.cpu(0))
frame = video[i].permute(2, 0, 1)
while True:
    time.sleep(1)
```

My guess would be `decord` is trying to decode the video in the background into the buffer and if you do not comsume them it'll just keep going. I don't believe this has anything to do with `Queue`.

**Comment by NevermindNilas** [comment]
I used your example with the only replacement being from `i` to `0` and I got this:
![image](https://github.com/python/cpython/assets/128264457/e9d28382-14fc-4117-94e6-b4230053ccde)

Decord does seem to utilize more memory than intended, this could be something related to how they optimize decoding. 

I went ahead and replaced decord with a subprocess call to ffmpeg just to make sure in the same code as suggested by you::
![image](https://github.com/python/cpython/assets/128264457/b9e76a11-baba-48c5-b0de-6b243a50875b)
The memory usage seemed to be fine,

I ran another quick test with my original code using again a subprocess call to ffmpeg and the memory usage was just fine:
![image](https://github.com/python/cpython/assets/128264457/307cc1d7-253b-41f2-9075-93cfdfff9939)

So I am guessing this is related to decord and not queue after all. Sorry for jumping to conclusions without more testing, I will forward this to the guys over there.

Thank you v much.
